### DISCUSSION TITLE
How might I create several concurrent reader connections in C++?

### DISCUSSION THREAD

**Discussion by chu-k** [duckdb/duckdb#14217] [Q&A] [answered]
The [Python examples ](https://duckdb.org/docs/guides/python/multiple_threads.html) show the ability to create several connections using `con.cursor()` and read simultaneously. 

In C++, I am creating new connections from a single DB instance as below:

```
std::shared_ptr<DuckDb> db = std::make_shared<duckdb::DuckDB>(path);

# instantiating several reader connections like...
Connection con(*db);
```

When I use execute a query in multiple readers and access the QueryResult, however, the one will be accessible while the rest fail with: {"exception_type":"Invalid Input","exception_message":"Attempting to execute an unsuccessful or closed pending query result"}

The Python docs seem to suggest this is possible. Am I creating multiple connections incorrectly? I don't believe I am sharing the same connection across threads as discussed [here](https://github.com/duckdb/duckdb/issues/11373). Each reader instance only has one connection, one active QueryResult.

**Comment by Tishj** [comment]
Your error suggests to me that you're using the same connection in all threads

**Reply by chu-k** [reply]
Hi Tishj, thanks for the reply. 

Yes that's what I have deduced, but is the way I'm instantiating these new connections incorrect in some way? I've also just tried removing the use of pointers altogether and passing the DB instance directly with the same result:

```
DuckDB db(path);

# and in the Reader class constructor

class Reader {
  Reader(DuckDB& db) : conn(db) {}
  ...
  Connection conn;
};

```

It seems like just calling [Connection constructor](https://github.com/duckdb/duckdb/blob/897f93d865c02d7fc2d01c5e3ee0ceec047b262d/src/main/connection.cpp#L21C1-L28C2) with the same DuckDB instance should handle adding new connections.

**Reply by Tishj** [reply, marked as answer]
I would have to see a full reproduction, your code looks correct to me, but if you're doing concurrent catalog modifications on the same entry (table/view/schema etc..) then it could just be that only one of them succeeds.

`Attempting to execute an unsuccessful or closed pending query result` sounds like you're missing a step of verification and calling Execute on a PendingQueryResult without first checking `HasError()`

**Reply by chu-k** [reply]
Led me down the right debugging path, much appreciated!
